File: carrotfarm/src/p2018_10_02/LinkedListCycle.java
package p2018_10_02;

public class LinkedListCycle {

	class ListNode {
		int val;
		ListNode next;
		ListNode(int x) {
			val = x;
			next = null;
		}
	}

	public boolean hasCycle(ListNode head) {

		boolean result = false;
		if (head == null) {
			return false;
		}

		ListNode rabbit = head;
		ListNode turtle = head;

		while(rabbit.next != null && rabbit.next.next != null) {

			turtle = turtle.next;
			rabbit = rabbit.next.next;
			if (turtle == rabbit) {
				result = true;
				break;
			}
		}
		return result;
	}

}

File: kangtaku/linkedListCycle.cpp
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    bool hasCycle(ListNode *head) {
        if (head ==  NULL) {
            return false;
        }
        
        ListNode* first = head;
        ListNode* second = head;
        
        while (first && first->next && second->next && second->next->next) {
            first = first->next;
            second = second->next->next;
            if (first == second) {
                return true;
            }
        }
        return false;
    }
};
